File: sdk/bootstrap/src/runner.ts
/**
 * The one door to the machine: every process this package starts, every file
 * it reads, goes through a {@link Runner}. The tests supply a scripted one and
 * assert on the argv that would have run; {@link processRunner} is the real one.
 *
 * The real runner carries the facts Foundry's WSL launcher learned the hard way
 * (docs/FROM-FOUNDRY-WSL-VLLM.md section 2), each with a test:
 *
 *   1. **wsl.exe's OWN output is UTF-16LE with a BOM; output from inside the
 *      distro is UTF-8, on the same handles.** `wsl -l -v` and every message
 *      wsl.exe writes itself ("there is no distribution with the supplied
 *      name") are UTF-16; a bash pipeline's stdout is UTF-8. The decode is
 *      decided per chunk by looking for interleaved NULs — {@link decodeWslBytes}.
 *
 *   2. **Every one-shot call gets its own timeout.** A booting distro or a
 *      blocking profile makes wsl.exe never return; `timeoutMs` is required on
 *      every call and a timeout is a reported failure, never a hang.
 *
 *   3. **Argument arrays, never shell strings.** `spawn(argv[0], argv.slice(1))`
 *      with `windowsHide`; nothing here is ever joined into a command line.
 *
 * The runner knows nothing about WSL beyond how to decode its bytes. What to
 * run, and how a Windows path or a backslash crosses into the guest, is
 * `wsl.ts`'s.
 */
import { spawn, spawnSync, type ChildProcess } from 'node:child_process';
import * as fs from 'node:fs';
import * as os from 'node:os';

export interface RunResult {
  /** Null when the process never started or was killed by the timeout. */
  code: number | null;
  stdout: string;
  stderr: string;
  /** Set when there is no exit code to report — a spawn error, or the timeout. */
  failure: string | null;
}

export interface RunOptions {
  /** Required. There is no call here that may block forever. */
  timeoutMs: number;
  /** Extra environment for the child, merged over the process's own. */
  env?: Readonly<Record<string, string>>;
}

export type OutputStream = 'stdout' | 'stderr';

export interface StreamOptions extends RunOptions {
  /** Called once per line, on either handle, as the lines arrive. */
  onLine: (line: string, stream: OutputStream) => void;
}

export interface Runner {
  readonly platform: NodeJS.Platform;
  readonly env: NodeJS.ProcessEnv;
  readonly homedir: string;
  /** Run `argv[0]` with `argv.slice(1)` and collect what it said. Never throws. */
  run(argv: readonly string[], options: RunOptions): Promise<RunResult>;
  /** The same, but `onLine` sees every line as it arrives. Never throws. */
  stream(argv: readonly string[], options: StreamOptions): Promise<RunResult>;
  fileExists(path: string): boolean;
  /** Throws on any failure; the caller names the refusal. */
  readFile(path: string): string;
  /** `fs.realpathSync.native`. Throws when the path cannot be resolved. */
  realpathNative(path: string): string;
}

/**
 * Bytes from a wsl.exe handle → text, whichever of the two encodings it is.
 *
 * The test is structural: UTF-16LE ASCII has a NUL after every character, and
 * UTF-8 text never contains a NUL at all. A BOM is stripped either way, because
 * `wsl -l -v`'s first line would otherwise begin with one and never match.
 */
export function decodeWslBytes(buffer: Buffer): string {
  const probe = buffer.subarray(0, Math.min(buffer.length, 256));
  const utf16 = probe.includes(0);
  return buffer.toString(utf16 ? 'utf16le' : 'utf8').replace(/^﻿/, '');
}

/**
 * Split text into lines on `\n`, `\r\n` AND a bare `\r`, because pip repaints
 * its progress bar with `\r` and no newline; splitting on both turns that into
 * successive lines rather than one line that never ends. Returns the complete
 * lines and the unterminated remainder.
 */
export function splitLines(pending: string): { lines: string[]; rest: string } {
  const parts = pending.split(/\r?\n|\r/);
  const rest = parts.pop() ?? '';
  return { lines: parts, rest };
}

/**
 * Kill a child and everything under it. `child.kill()` reaches wsl.exe alone
 * and leaves the guest-side process it is relaying for; taskkill's tree flag is
 * the only thing on Windows that gets the rest. Used for timed-out calls and
 * nothing else — never against a service, which is stopped by its supervisor.
 */
function killTree(child: ChildProcess, platform: NodeJS.Platform): void {
  if (child.pid === undefined || child.exitCode !== null) return;
  if (platform === 'win32') {
    try {
      spawnSync('taskkill', ['/pid', String(child.pid), '/t', '/f'], { windowsHide: true });
      return;
    } catch {
      // fall through to the signal
    }
  }
  try {
    child.kill('SIGTERM');
  } catch {
    // already gone
  }
}

function launch(
  argv: readonly string[],
  options: RunOptions,
  onChunk: (chunk: Buffer, stream: OutputStream) => void,
  onFinish: (result: RunResult) => void,
): void {
  const program = argv[0];
  if (program === undefined) {
    onFinish({ code: null, stdout: '', stderr: '', failure: 'empty argv' });
    return;
  }
  let child: ChildProcess;
  try {
    child = spawn(program, argv.slice(1), {
      windowsHide: true,
      env: options.env === undefined ? process.env : { ...process.env, ...options.env },
    });
  } catch (err) {
    onFinish({ code: null, stdout: '', stderr: '', failure: (err as Error).message });
    return;
  }

  const out: Buffer[] = [];
  const err: Buffer[] = [];
  let settled = false;
  const finish = (code: number | null, failure: string | null): void => {
    if (settled) return;
    settled = true;
    clearTimeout(timer);
    onFinish({
      code,
      stdout: decodeWslBytes(Buffer.concat(out)),
      stderr: decodeWslBytes(Buffer.concat(err)),
      failure,
    });
  };
  const timer = setTimeout(() => {
    killTree(child, process.platform);
    finish(null, `${program} did not answer within ${Math.round(options.timeoutMs / 1000)}s`);
  }, options.timeoutMs);

  child.stdout?.on('data', (chunk: Buffer) => {
    out.push(chunk);
    onChunk(chunk, 'stdout');
  });
  child.stderr?.on('data', (chunk: Buffer) => {
    err.push(chunk);
    onChunk(chunk, 'stderr');
  });
  child.on('error', (e) => finish(null, e.message));
  child.on('close', (code) => finish(code, null));
}

/** The real machine. */
export function processRunner(): Runner {
  return {
    platform: process.platform,
    env: process.env,
    homedir: os.homedir(),
    run(argv, options) {
      return new Promise<RunResult>((resolve) => {
        launch(argv, options, () => undefined, resolve);
      });
    },
    stream(argv, options) {
      return new Promise<RunResult>((resolve) => {
        const pending: Record<OutputStream, string> = { stdout: '', stderr: '' };
        launch(
          argv,
          options,
          (chunk, stream) => {
            // Decoded PER CHUNK: a UTF-16 message from wsl.exe itself can arrive
            // between two UTF-8 pip lines on the same handle.
            const { lines, rest } = splitLines(pending[stream] + decodeWslBytes(chunk));
            pending[stream] = rest;
            for (const line of lines) if (line.trim().length > 0) options.onLine(line, stream);
          },
          (result) => {
            for (const stream of ['stdout', 'stderr'] as const) {
              if (pending[stream].trim().length > 0) options.onLine(pending[stream], stream);
            }
            resolve(result);
          },
        );
      });
    },
    fileExists: (path) => fs.existsSync(path),
    readFile: (path) => fs.readFileSync(path, 'utf-8'),
    realpathNative: (path) => fs.realpathSync.native(path),
  };
}

File: crucible/host/runner.py
"""The one door to a subprocess and to the engine's `/v1/ping`.

Everything the host does to the machine goes through a `Runner`, for the reason
`sdk/bootstrap/src/runner.ts` gives on the other side of this phase: a test
supplies a scripted one and asserts on the ARGV that would have run, so the boot
recipe, the recovery recipes and the whole install sequence are exercised on a
machine that is not Windows and has no `crucible` distro.

Three rules carried over from that file, each of which was a defect somewhere:

1. **Argument arrays, never shell strings.** Nothing here is joined into a
   command line. (BookForge's memory `wsl-exe-implicit-shell-trap.md`: wsl.exe
   PRE-EXPANDS `$var` unless `--exec`.)
2. **Every call has a timeout.** A booting distro or a blocking profile makes
   `wsl.exe` never return, and a tray that hangs is a tray with no menu.
3. **A failure is a RESULT, not an exception.** The caller decides what a
   non-zero exit means; several of them mean "not yet" rather than "wrong".

And a fourth this file learned on its own, deploy 1.0.4:

4. **The pipes are BYTES, and this module decides what they say.** `wsl.exe`
   writes its own diagnostics as UTF-16LE; see `_decode_pipe`.
"""

from __future__ import annotations

import codecs
import subprocess
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Mapping, Protocol, Sequence


@dataclass(frozen=True)
class RunResult:
    """What a command said. `code` is None when it never produced one."""

    code: int | None
    stdout: str
    stderr: str
    #: Set when there is no exit code to report: a spawn error, or the timeout.
    failure: str | None

    @property
    def ok(self) -> bool:
        return self.failure is None and self.code == 0

    def said(self) -> str:
        """The most useful line to put in a log or a refusal."""
        for candidate in (self.stderr.strip(), self.stdout.strip(), self.failure):
            if candidate:
                return candidate[:400]
        return f"exit {self.code}"


class Runner(Protocol):
    """What the host is allowed to do to the machine."""

    @property
    def platform(self) -> str: ...

    @property
    def env(self) -> Mapping[str, str]: ...

    def run(
        self,
        argv: Sequence[str],
        *,
        timeout_s: float,
        env: Mapping[str, str] | None = None,
    ) -> RunResult:
        """Run and collect. Never raises for a non-zero exit or a timeout."""
        ...

    def get(self, url: str, *, timeout_s: float) -> int | None:
        """The HTTP status of a GET, or None when nothing answered."""
        ...

    def spawn(
        self,
        argv: Sequence[str],
        *,
        env: Mapping[str, str] | None = None,
    ) -> "Child":
        """Start a long-lived child (the host-mode server) and return a handle."""
        ...


class Child(Protocol):
    """A process the host started and is responsible for."""

    @property
    def pid(self) -> int: ...

    def poll(self) -> int | None:
        """The exit code, or None while it is still running."""
        ...

    def terminate(self) -> None: ...

    def wait(self, timeout_s: float) -> int | None: ...


class ControlledChild:
    """An owned engine exits through its lifespan, never TerminateProcess."""

    def __init__(self, process: subprocess.Popen) -> None:
        self._process = process

    @property
    def pid(self) -> int:
        return self._process.pid

    def poll(self) -> int | None:
        return self._process.poll()

    def terminate(self) -> None:
        # Idempotent; closing the writer also handles an engine already exiting.
        if self._process.stdin is not None:
            self._process.stdin.close()

    def wait(self, timeout_s: float) -> int:
        return self._process.wait(timeout=timeout_s)


class ProcessRunner:
    """The real one. `subprocess` plus `urllib`, and nothing else."""

    def __init__(
        self, platform: str, env: Mapping[str, str], cwd: str | None = None
    ) -> None:
        self._platform = platform
        self._env = dict(env)
        #: WHERE CHILDREN START, and it is not cosmetic. A child inherits this
        #: process's working directory, and the orchestrator's is inside its own
        #: installation (installation.json records
        #: `...\Crucible\host\Lib\site-packages`, deliberately, so that
        #: `-m crucible.cli` imports). A `wsl.exe` child therefore holds a handle
        #: on `Crucible\host` — and keeps holding it after the orchestrator
        #: exits, which is what stopped an upgrade on 2026-09-16 with
        #: "Move-Item: the process cannot access the file because it is being
        #: used by another process", naming nothing. Sysinternals `handle64`
        #: found two orphaned wsl.exe and a wslhost.exe on that directory.
        #:
        #: The caller passes CRUCIBLE_HOME: the server's own state directory,
        #: which is what the systemd unit uses as WorkingDirectory for the same
        #: reason, and which the installer never moves. NOT the user's home —
        #: `console_script` records the ImportError that follows from a working
        #: directory landing on sys.path.
        self._cwd = cwd

    @property
    def platform(self) -> str:
        return self._platform

    @property
    def env(self) -> Mapping[str, str]:
        return self._env

    def _child_env(self, env: Mapping[str, str] | None) -> dict[str, str] | None:
        if env is None:
            return None
        merged = dict(self._env)
        merged.update(env)
        return merged

    def run(
        self,
        argv: Sequence[str],
        *,
        timeout_s: float,
        env: Mapping[str, str] | None = None,
    ) -> RunResult:
        try:
            completed = subprocess.run(
                list(argv),
                capture_output=True,
                timeout=timeout_s,
                env=self._child_env(env),
                cwd=self._cwd,
                # A tray program has no console; a child that opens one is a
                # window flashing on somebody's desktop every fifteen seconds.
                creationflags=_no_window_flag(self._platform),
                # NOT `text=True`. Asking subprocess to decode means asking it
                # to decode with the LOCALE codec, and one of the two programs
                # on the other end of this pipe does not use it — `_decode_pipe`.
            )
        except subprocess.TimeoutExpired:
            return RunResult(
                code=None,
                stdout="",
                stderr="",
                failure=f"timed out after {timeout_s:.0f}s",
            )
        except OSError as exc:
            return RunResult(code=None, stdout="", stderr="", failure=str(exc))
        return RunResult(
            code=completed.returncode,
            stdout=_decode_pipe(completed.stdout),
            stderr=_decode_pipe(completed.stderr),
            failure=None,
        )

    def get(self, url: str, *, timeout_s: float) -> int | None:
        try:
            with urllib.request.urlopen(url, timeout=timeout_s) as response:
                return int(response.status)
        except urllib.error.HTTPError as exc:
            # It ANSWERED. 401 from `/v1/ping` would still mean a server is
            # there, and "there is a server and it refused me" is not the same
            # fact as "nothing is listening" — see `presence.py`.
            return int(exc.code)
        except (urllib.error.URLError, OSError, ValueError):
            return None

    def spawn(
        self,
        argv: Sequence[str],
        *,
        env: Mapping[str, str] | None = None,
    ) -> Child:
        controlled = list(argv)[1:] == ["-m", "crucible.cli", "serve", "--controller-stdin"]
        child = subprocess.Popen(
            list(argv),
            env=self._child_env(env),
            cwd=self._cwd,
            creationflags=_no_window_flag(self._platform),
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            stdin=subprocess.PIPE if controlled else subprocess.DEVNULL,
        )
        return ControlledChild(child) if controlled else child  # type: ignore[return-value]



#: How many bytes of a stream are enough to recognise UTF-16LE by its shape.
#: A whole `wsl -l -v` table would do as well; the point of a window is that
#: the guest's own output, which can be megabytes, is not walked twice.
_UTF16_SNIFF_BYTES = 64


def _decode_pipe(raw: bytes | None) -> str:
    """Turn one captured pipe into text, and MEASURE which codec wrote it.

    `wsl.exe` IS THE ONLY TOOL THIS HOST RUNS THAT NEEDS THIS, and the reason
    is that it is really two programs. Its OWN messages — "There is no
    distribution with the supplied name.", "Error code: Wsl/Service/
    WSL_E_DISTRO_NOT_FOUND", the `wsl -l -v` table — are written by the Windows
    side as UTF-16LE, which is what the Windows console API takes. Everything
    it `--exec`s is a program inside the guest, and its bytes are RELAYED, so
    they arrive exactly as Linux wrote them: UTF-8. One command can therefore
    produce a UTF-16 stderr and a UTF-8 stdout, which is why this decides per
    stream and not once per call.

    MEASURED 2026-09-19, deploy 1.0.4. The runner asked `subprocess.run` for
    text, so both streams were decoded with the locale codec — under which
    every NUL of a UTF-16LE string is a perfectly good character — and the one
    sentence naming why the deploy had failed reached `host.log` as
    `T\\x00h\\x00e\\x00r\\x00e\\x00 \\x00i\\x00s\\x00 …`. Nothing was lost; it
    was simply unreadable, which for a diagnostic is the same thing.

    Two facts identify it, both from the bytes themselves rather than from the
    argv, because a runner that decided by command name would be a second
    owner of the question "what is wsl.exe": a UTF-16LE byte-order mark, and —
    for the streams that carry none, which is what 1.0.4 measured — a high
    byte of zero under every ASCII character in the opening window.

    `errors="replace"` and never `"ignore"`: a byte nothing can decode becomes
    U+FFFD, a character a person reading the log can SEE, rather than a hole in
    a sentence that reads as if it were complete.
    """
    if not raw:
        return ""
    if raw.startswith(codecs.BOM_UTF16_LE):
        return _newlines(raw[len(codecs.BOM_UTF16_LE) :].decode("utf-16-le", errors="replace"))
    window = raw[:_UTF16_SNIFF_BYTES]
    if len(window) >= 2 and all(window[i] == 0 for i in range(1, len(window), 2)):
        return _newlines(raw.decode("utf-16-le", errors="replace"))
    return _newlines(raw.decode("utf-8", errors="replace"))


def _newlines(text: str) -> str:
    """CRLF and CR to LF — what `text=True` used to do on the way past.

    Not cosmetic and not new behaviour: `parse_wsl_list` and every other reader
    in this package was written against universal-newline output, and wsl.exe
    is a Windows program that ends its lines the Windows way.
    """
    return text.replace("\r\n", "\n").replace("\r", "\n")


def _no_window_flag(platform: str) -> int:
    """`CREATE_NO_WINDOW` on Windows, 0 elsewhere.

    Read off `subprocess` rather than written as `0x08000000`, but only when
    the attribute is there: the constant does not exist on Linux, and this
    module is imported by the test suite, which runs in WSL.
    """
    if platform != "win32":
        return 0
    return int(getattr(subprocess, "CREATE_NO_WINDOW", 0))
